comp/systray/systray/systray.go

```diff
@@ -0,0 +1,414 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+//go:build windows
+// +build windows
+
+package systray
+
+//#include "uac.h"
+import "C"
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"runtime"
+	"strings"
+	"sync"
+	"time"
+	"unsafe"
+
+	"github.com/DataDog/datadog-agent/comp/core/config"
+	"github.com/DataDog/datadog-agent/comp/core/flare"
+	"github.com/DataDog/datadog-agent/comp/core/log"
+	"github.com/DataDog/datadog-agent/comp/systray/internal"
+	pkglog "github.com/DataDog/datadog-agent/pkg/util/log"
+	"github.com/DataDog/datadog-agent/pkg/util/winutil"
+	"github.com/DataDog/datadog-agent/pkg/version"
+
+	"github.com/lxn/walk"
+	"github.com/lxn/win"
+	"go.uber.org/fx"
+	"golang.org/x/sys/windows"
+)
+
+type dependencies struct {
+	fx.In
+
+	Lc         fx.Lifecycle
+	Shutdowner fx.Shutdowner
+
+	Log    log.Component
+	Config config.Component
+	Flare  flare.Component
+	Params internal.BundleParams
+}
+
+type systray struct {
+	// For triggering Shutdown
+	shutdowner fx.Shutdowner
+
+	log    log.Component
+	config config.Component
+	flare  flare.Component
+	params internal.BundleParams
+
+	isUserAdmin bool
+
+	// allocated in start, destroyed in stop
+	singletonEventHandle windows.Handle
+
+	// Window management
+	notifyWindowToStop func()
+	routineWaitGroup   sync.WaitGroup
+}
+
+type menuItem struct {
+	label   string
+	handler walk.EventHandler
+	enabled bool
+}
+
+const (
+	launchGraceTime       = 2
+	eventname             = "ddtray-event"
+	cmdTextStartService   = "StartService"
+	cmdTextStopService    = "StopService"
+	cmdTextRestartService = "RestartService"
+	cmdTextConfig         = "Config"
+	menuSeparator         = "SEPARATOR"
+)
+
+var (
+	cmds = map[string]func(*systray){
+		cmdTextStartService:   onStart,
+		cmdTextStopService:    onStop,
+		cmdTextRestartService: onRestart,
+		cmdTextConfig:         onConfigure,
+	}
+)
+
+// newSystray creates a new systray component, which will start and stop based on
+// the fx Lifecycle
+func newSystray(deps dependencies) (Component, error) {
+	// fx init
+	s := &systray{
+		log:        deps.Log,
+		config:     deps.Config,
+		flare:      deps.Flare,
+		params:     deps.Params,
+		shutdowner: deps.Shutdowner,
+	}
+
+	// fx lifecycle hooks
+	deps.Lc.Append(fx.Hook{OnStart: s.start, OnStop: s.stop})
+
+	// init vars
+	isAdmin, err := winutil.IsUserAnAdmin()
+	if err != nil {
+		s.log.Warnf("Failed to call IsUserAnAdmin %v", err)
+		// If we cannot determine if the user is admin or not let the user allow to click on the buttons.
+		s.isUserAdmin = true
+	} else {
+		s.isUserAdmin = isAdmin
+	}
+
+	return s, nil
```

Comment:
Note: If there is no error, the returned type of `newSystray` can just be `Component`.